sql-odbc/src/PowerBIConnector/OdfeSqlOdbcPBIConnector.pq

```diff
@@ -1,42 +1,203 @@
+        Server = 
+            if Port <> null then
+            [
+                Host = Host,
+                Port = Port
+            ]
+            else
+            [
+                Host = Host
+            ],
+
         ConnectionString = [
-            Driver = "ODFE SQL ODBC Driver",
-            Host = Host,
-            Port = Port
+            Driver = "ODFE SQL ODBC Driver"
         ],
-        OdbcDatasource = Odbc.DataSource(ConnectionString)
+
+        SQLGetTypeInfo = (types) => 
+            if (EnableTraceOutput <> true) then types else
+            let
+                // Outputting the entire table might be too large, and result in the value being truncated.
+                // We can output a row at a time instead with Table.TransformRows()
+                rows = Table.TransformRows(types, each Diagnostics.LogValue("SQLGetTypeInfo " & _[TYPE_NAME], _)),
+                toTable = Table.FromRecords(rows)
+            in
+                Value.ReplaceType(toTable, Value.Type(types)),
+
+        // SQLColumns is a function handler that receives the results of an ODBC call to SQLColumns().
+        SQLColumns = (catalogName, schemaName, tableName, columnName, source) =>
+            if (EnableTraceOutput <> true) then source else
+            // the if statement conditions will force the values to evaluated/written to diagnostics
+            if (Diagnostics.LogValue("SQLColumns.TableName", tableName) <> "***" and Diagnostics.LogValue("SQLColumns.ColumnName", columnName) <> "***") then
+                let
+                    // Outputting the entire table might be too large, and result in the value being truncated.
+                    // We can output a row at a time instead with Table.TransformRows()
+                    rows = Table.TransformRows(source, each Diagnostics.LogValue("SQLColumns", _)),
+                    toTable = Table.FromRecords(rows)
+                in
+                    Value.ReplaceType(toTable, Value.Type(source))
+            else
+                source,
+             
+        // Add support for `LIMIT` and `OFFSET` clauses (rather than `TOP`)
+        AstVisitor = [
+            // format is "LIMIT [<skip>,]<take>" - ex. LIMIT 2,10 or LIMIT 10
+            LimitClause = (skip, take) =>
+                if (take = null) then
+                    ...
+                else
+                    let
+                        skip =
+                            if (skip = null or skip = 0) then
+                                ""
+                            else
+                                Number.ToText(skip) & ","
+                    in
+                        [
+                            Text = Text.Format("LIMIT #{0}#{1}", { skip, take }),
+                            Location = "AfterQuerySpecification"
+                        ]
+        ],
+
+        OdbcDatasource = Odbc.DataSource(ConnectionString & Server & CredentialConnectionString, [
```

Comment:
Shouldn't this concatenate EncryptedConnectionString too?

Reply:
added